### PR TITLE
Fix digest authentication on SocketsHttpHandler

### PR DESCRIPTION
**Pull request by MarcoRossignoli**
fixes https://github.com/dotnet/corefx/issues/37729

Problem I found:

1) Updated code to follow https://tools.ietf.org/html/rfc2617#section-3.2.2

```
qop
     Indicates what "quality of protection" the client has applied to
     the message. If present, its value MUST be one of the alternatives
     the server indicated it supports in the WWW-Authenticate header.
     These values affect the computation of the request-digest. Note
     that this is a single token, not a quoted list of alternatives as
     in WWW- Authenticate.  This directive is optional in order to
     preserve backward compatibility with a minimal implementation of
     RFC 2069 [6], but SHOULD be used if the server indicated that qop
     is supported by providing a qop directive in the WWW-Authenticate
     header field.
```
At the moment `qop` is always included and not follow `but SHOULD be used if the server indicated that qop is supported by providing a qop directive in the WWW-Authenticate header field`

2) Updated code to follow: https://tools.ietf.org/html/rfc2617#section-3.2.2

```
The values of the opaque and algorithm fields must be those supplied
   in the WWW-Authenticate response header for the entity being
   requested.
```
To be compatible with .NET full and WinHttpHandler we'll send `algorithm` only if specified in `WWW-Authenticate` header, otherwise we "silently" use default.

3) Fixed `-sess` tests, passing `qop` otherwise now that we follow 1) server side validation fails because we don't pass nouces anymore. 
    But it's not correct avoid to pass `qop`, `-sess` is present only in rfc2617 where `qop` `it SHOULD be used by all implementations compliant with this version of the Digest scheme.`
    rfc2069 does not mention this kind of algorithm.

4) Implemented missing part https://tools.ietf.org/html/rfc2617#section-3.2.2.1 (some tests fail without)
```
If the "qop" directive is not present (this construction is for
   compatibility with RFC 2069):

      request-digest  =
                 <"> < KD ( H(A1), unq(nonce-value) ":" H(A2) ) >
   <">
```

I added tests where I validate the presence of fields using regex.

I did issue's test on fiddler, log:
```
GET https://jigsaw.w3.org/HTTP/Digest/ HTTP/1.1
Host: jigsaw.w3.org

HTTP/1.1 401 Unauthorized
Date: Tue, 04 Jun 2019 20:11:58 GMT
Content-Length: 261
Content-Type: text/html;charset=ISO-8859-1
Server: Jigsaw/2.3.0-beta2
WWW-Authenticate: Digest realm="test", domain="/HTTP/Digest", nonce="b9aa1067c188f78cdbedd405741231c2"
Strict-Transport-Security: max-age=15552015; includeSubDomains; preload
Public-Key-Pins: pin-sha256="cN0QSpPIkuwpT6iP2YjEo1bEwGpH/yiUn6yhdy+HNto="; pin-sha256="WGJkyYjx1QMdMe0UqlyOKXtydPDVrk7sl2fV+nNm1r4="; pin-sha256="LrKdTxZLRTvyHM4/atX2nquX9BeHRZMCxg3cf4rhc2I="; max-age=864000
X-Frame-Options: deny
X-XSS-Protection: 1; mode=block

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
               "http://www.w3.org/TR/html4/loose.dtd">
<html>
<head>
 <title>Unauthorized</title>
</head>
<body>
<h1>Unauthorized access</h1><p>You are denied access to this resource.</body>
</html>
```
```
GET https://jigsaw.w3.org/HTTP/Digest/ HTTP/1.1
Host: jigsaw.w3.org
Authorization: Digest username="guest", realm="test", nonce="b9aa1067c188f78cdbedd405741231c2", uri="/HTTP/Digest/", response="5641fcab8f7b107e90eab9fc059008fd"

HTTP/1.1 200 OK
Date: Tue, 04 Jun 2019 20:11:58 GMT
Content-Length: 459
Content-Location: https://jigsaw.w3.org/HTTP/Digest/ok.html
Content-Type: text/html
Etag: "6km9jn:qogunjao"
Last-Modified: Thu, 01 Jun 2000 22:07:10 GMT
Server: Jigsaw/2.3.0-beta2
Strict-Transport-Security: max-age=15552015; includeSubDomains; preload
Public-Key-Pins: pin-sha256="cN0QSpPIkuwpT6iP2YjEo1bEwGpH/yiUn6yhdy+HNto="; pin-sha256="WGJkyYjx1QMdMe0UqlyOKXtydPDVrk7sl2fV+nNm1r4="; pin-sha256="LrKdTxZLRTvyHM4/atX2nquX9BeHRZMCxg3cf4rhc2I="; max-age=864000
X-Frame-Options: deny
X-XSS-Protection: 1; mode=block

<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML//EN">
<html>
  <head>
    <title>Digest Authentication test page</title>
<!-- Changed by: Yves Lafon, 22-Feb-1999 -->
  </head>

<BODY BGCOLOR="white">
<P>
      <A HREF=".."><IMG SRC="/icons/jigsaw" ALT="Jigsaw" BORDER="0" WIDTH="212"
	  HEIGHT="49"></A><IMG SRC="/icons/jigpower.gif" WIDTH="94" HEIGHT="52" ALT="Jigsaw Powered !"
	BORDER="0" ALIGN="Right">
      
<P>
<HR>
<P>Your browser made it!
  </body>
</html>
```

cc area owners: @geoffkizer, @wfurt, @davidsh, @karelz

### CODE REVIEW

**Comment by wfurt** [comment]
MacOS failure should be fixed by #38228

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/AuthenticationHelper.Digest.cs [hunk 1]
```diff
@@ -47,7 +47,8 @@ public static async Task<string> GetDigestTokenForCredential(NetworkCredential c
             // It is mandatory for servers to implement sha-256 per RFC 7616
             // Keep MD5 for backward compatibility.
             string algorithm;
-            if (digestResponse.Parameters.TryGetValue(Algorithm, out algorithm))
+            bool isAlgorithmSpecified;
+            if ((isAlgorithmSpecified = digestResponse.Parameters.TryGetValue(Algorithm, out algorithm)))
```
Comment (wfurt) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/AuthenticationHelper.Digest.cs]: can you please put the assignment before `if`
```c#
bool isAlgorithmSpecified = digestResponse.Parameters(..);
if (isAlgorithmSpecified)
{
```

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/AuthenticationHelper.Digest.cs [hunk 2]
```diff
@@ -114,7 +115,8 @@ public static async Task<string> GetDigestTokenForCredential(NetworkCredential c
 
             // Set qop, default is auth
             string qop = Auth;
-            if (digestResponse.Parameters.ContainsKey(Qop))
+            bool isQopSpecified;
+            if ((isQopSpecified = digestResponse.Parameters.ContainsKey(Qop)))
```
Comment (wfurt) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/AuthenticationHelper.Digest.cs]: same as above.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/AuthenticationHelper.Digest.cs (hunk 1 above)
Comment (MarcoRossignoli) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/AuthenticationHelper.Digest.cs]: done

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/AuthenticationHelper.Digest.cs (hunk 2 above)
Comment (MarcoRossignoli) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/AuthenticationHelper.Digest.cs]: done

**Comment by davidsh** [comment]
CI Failures not related to PR.

**Pull request merged**